File: app/src/main/java/com/lxy/molweightcalculator/parsing/ErrorCode.java
package com.lxy.molweightcalculator.parsing;

import androidx.annotation.IntDef;

@IntDef({ErrorCode.NO_ERROR,
        ErrorCode.EMPTY_FORMULA,
        ErrorCode.NO_ELEMENT,
        ErrorCode.MISMATCHED_BRACKET,
        ErrorCode.INVALID_TOKEN,
        ErrorCode.INVALID_ELEMENT,
        ErrorCode.ELEMENT_COUNT_TOO_LARGE,
        ErrorCode.ELEMENT_COUNT_OVERFLOW,
        ErrorCode.WEIGHT_OVERFLOW,
})
public @interface ErrorCode {
    int NO_ERROR = 0;
    int EMPTY_FORMULA = 1;
    int NO_ELEMENT = 2;
    int MISMATCHED_BRACKET = 3;
    int INVALID_TOKEN = 4;
    int INVALID_ELEMENT = 5;
    int ELEMENT_COUNT_TOO_LARGE = 6;
    int ELEMENT_COUNT_OVERFLOW = 7;
    int WEIGHT_OVERFLOW = 8;
    int MINIMUM = EMPTY_FORMULA;
    int MAXIMUM = WEIGHT_OVERFLOW;
}

File: app/src/main/java/com/lxy/molweightcalculator/parsing/ErrorCode.kt
package com.lxy.molweightcalculator.parsing

enum class ErrorCode {
    EmptyFormula,
    NoElement,
    MismatchedBracket,
    InvalidToken,
    InvalidElement,
    ElementCountTooLarge,
    ElementCountOverflow,
    WeightOverflow,
    FormulaTooLong;

    companion object {
        private val NO_START_END_MASK = (1 shl EmptyFormula.ordinal) or
                (1 shl ElementCountOverflow.ordinal) or
                (1 shl WeightOverflow.ordinal) or
                (1 shl FormulaTooLong.ordinal)

        fun valueOf(value: Int): ErrorCode = entries[value]
    }

    val isInvalidBracket get() = this == MismatchedBracket

    val hasStartEnd get() = ((1 shl ordinal) and NO_START_END_MASK) == 0
}